sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -249,6 +249,20 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
               s"join condition '${condition.sql}' " +
                 s"of type ${condition.dataType.catalogString} is not a boolean.")
 
+          case j @ AsOfJoin(_, _, _, Some(condition), _, _, _)
+              if condition.dataType != BooleanType =>
+            failAnalysis(
+              s"join condition '${condition.sql}' " +
+                s"of type ${condition.dataType.catalogString} is not a boolean.")
+
+          case j @ AsOfJoin(_, _, _, _, _, _, Some(toleranceAssertion)) =>
+            if (!toleranceAssertion.foldable) {
+              failAnalysis("Input argument tolerance must be a constant.")
+            }
+            if (!toleranceAssertion.eval().asInstanceOf[Boolean]) {
```

Comment:
I think `.asInstanceOf[Number].longValue >= 0` is clearer.

Reply:
That makes sense, but the result of `eval()` is type dependent, e.g. if the `tolerance` is `CatelendarInterval`, the cast will fail, IIUC, whereas if we make the condition with expressions, it will only return `Boolean` value after the analysis phase. WDYT?

Reply:
ah that's why it's called `toleranceAssertion`, it's a predicate!